velox/exec/GroupingSet.cpp

```diff
@@ -1081,6 +1094,48 @@ bool GroupingSet::mergeNext(
   VELOX_UNREACHABLE();
 }
 
+bool GroupingSet::mergeNextForDistinct(
+    int32_t maxOutputRows,
+    int32_t maxOutputBytes,
+    const RowVectorPtr& result) {
+  VELOX_CHECK(isDistinct());
+
+  bool nextKeyIsEqual{false};
+  bool newDistinct{true};
+  int32_t numOutputRows{0};
+  for (;;) {
+    const auto next = merge_->nextWithEquals();
+    auto* stream = next.first;
+    if (stream == nullptr) {
+      result->resize(numOutputRows);
+      break;
+    }
+    if (stream->id() == 0) {
+      // A spilled row is not distinct if it is contained in the first stream
+      // which has already been output as distinct before we triggered the
+      // spilling.
+      newDistinct = false;
+    }
+    nextKeyIsEqual = next.second;
+    if (!nextKeyIsEqual) {
+      if (newDistinct) {
+        // yield result.
+        result->copy(
+            &stream->current(), numOutputRows++, stream->currentIndex(), 1);
+      }
+      newDistinct = true;
+    }
+    stream->pop();
+    if (!nextKeyIsEqual &&
+        ((numOutputRows >= maxOutputRows) ||
+         (result->estimateFlatSize() >= maxOutputBytes))) {
```

Comment:
Do we want to estimateFlatSize() after each row? Can we estimate number of output rows before the loop?